docs/design/scheduler_extender.md

```diff
@@ -0,0 +1,116 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/design/scheduler_extender.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Scheduler extender
+
+There are three ways to add new scheduling rules (predicates and priority functions) to Kubernetes: (1) by adding these rules to the scheduler and recompiling (described here: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/scheduler.md), (2) implementing your own scheduler process that runs instead of, or alongside of, the standard Kubernetes scheduler, (3) implementing a "scheduler extender" process that the standard Kubernetes scheduler calls out to as a final pass when making scheduling decisions. This document describes the third approach. (Note that the three approaches are not mutually exclusive.)
+
+When scheduling a pod, the extender allows an external process to filter and prioritize nodes. Two separate http/https calls are issued to the extender, one for "filter" and one for "prioritize" actions. To use the extender, you must create a scheduler policy configuration file. The configuration specifies how to reach the extender, whether to use http or https and the timeout. Extender timeouts are treated as scheduling errors and the pod remains unscheduled.
```

Comment:
@davidopp Done
@timothysc 
Sorry if this looks a little long, but here you go.

Its not about running one or many schedulers. Scheduling should be based on all resources in the cluster. The base k8s scheduler does a good job of scheduling based on cpu/mem/affinity etc. What is left out is storage and network resource based scheduling. When storage is remote (like GCEPersistentDisk, EBS etc), its not a scheduling constraint. But if you are on a hypercoverged infrastructure, it is absolutely. Placing workloads local/close to storage does wonders for performance. Similarly, network bandwidth is not an infinite resource on a server. When scheduling a pod (irrespective of which mechanism is used to achieve this), all of these are important requirements for performance sensitive applications.

Now lets talk about the mechanisms. As the document says, there are three mechanisms to do this. I will quote that verbtim here.
```
There are three ways to add new scheduling rules (predicates and priority functions) to Kubernetes: (1) by adding these rules to the scheduler and recompiling (described here: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/scheduler.md), (2) implementing your own scheduler process that runs instead of, or alongside of, the standard Kubernetes scheduler, (3) implementing a "scheduler extender" process that the standard Kubernetes scheduler calls out to as a final pass when making scheduling decisions. This document describes the third approach. (Note that the three approaches are not mutually exclusive.)
```
You are very familiar with Option 1. #17197 is way of accomplishing Option 2. So why is there a need for a third method? What I have described above doesn't fit in Options 1 or 2. In Option 1, k8s scheduler is not aware of all storage, network resources in the cluster. In Option 2, responsibility is split up at a pod level. What Option 3 provides is for the responsibility to be split up at a resource level. k8s schedules based on the resources it knows and the extenders schedule based on the resources they know and manage. Makes sense?